backend ready for check

Pancfly opened the pull request.

Pancfly committed:
backend ready for check

Boortcore commented:
# Резюме:
## Backend

### Критерии, влияющие на работоспособность
- [x] <!--3.64--> В репозитории есть все необходимые инфраструктурные файлы:
    - [x] в проекте есть `package.json`;
    - [x] `.editorconfig`;
    - [x] `.eslintrc`, расширяющий конфигурацию airbnb-base, а также необходимые для работы линтера dev-зависимости;
    - [x] `.gitignore`;
    - [x] в `.eslintrc` добавлено исключение для `_id`. Директивы `eslint-disable`, `eslint-disable-line` и `eslint-disable-next-line` запрещены.
- [x] <!--3.64--> Нет ошибок линтера. При выполнении команды `npx eslint .` они должны отсутствовать. 
- [x] <!--3.64--> В разделе scripts файла `package.json` есть команды:
    - [x] `npm run start`, которая запускает сервер на `localhost:3000`;
    - [x] `npm run dev`, которая запускает сервер на `localhost:3000` с хот релоудом.
- [x] <!--21.81--> Описанные роуты работают корректно: 
    - [x] запрос на `GET /users/me` возвращает информацию о пользователе (email и имя);
    - [x] `PATCH /users/me` — обновляет информацию о пользователе;
    - [x] `GET /movies` — все сохранённые пользователем фильмы;
    - [x] `POST /movies` — создаёт фильм с переданными в теле данными;
    - [x] `DELETE /movies/:id` — удаляет сохранённый фильм по `_id`;
    - [x] `POST /signup` — создаёт пользователя с переданными в теле данными;
    - [x] `POST /signin` — возвращает JWT, если в теле запроса переданы правильные почта и пароль.
    - [x] Если вы сохраняете JWT в куках, роут `/signout` должен удалять JWT из куки.
- [x] <!--3.64--> Пользователь не может удалить сохранённый фильм из профиля другого пользователя. 
- [x] <!--3.64--> Все роуты, кроме `/signin` и `/signup`, защищены авторизацией.
- [x] <!--3.64--> Роуты пользователей и роуты фильмов описаны в отдельных файлах.
- [x] <!--7.28--> Ошибки API обрабатываются: 
    - [x] используются статус-коды ошибок: 400, 401, 403, 404, 409, 500;
    - [x] если в запросе что-то не так, сервер возвращает статус ответа, который соответствует типу ошибки;
_- При обновлении данных пользователя с использованием почтового ящика, который принадлежит другому юзеру, необходимо возвращать ошибку 409._
_Сейчас возвращается ошибка 500._
    - [x] сообщение об ошибке соответствует ее типу;
    - [x] асинхронные обработчики завершены блоком `catch`;
    - [x] API не возвращает стандартных ошибок базы данных или Node.js.
- [x] <!--3.64--> Реализовано бережное хранение пароля:
    - [x] пароль хранится в виде хеша;
    - [x] API не возвращает хеш пароля клиенту.
- [x] <!--7.28--> Валидация реализована верно:
    - [x] Запрос валидируется перед передачей контроллеру. Тело и, где необходимо, заголовки и параметры проверяются по соответствующим схемам. Если запрос не соответствует схеме, обработка не передаётся контроллеру, и клиент получает ошибку валидации.
    - [x] Данные валидируются перед добавлением в базу.
- [x] <!--7.28--> В production-режиме адрес базы данных берётся из `process.env`.
_- Адрес прописан статически и не зависит от режима запуска приложения._
- [x] <!--3.64--> К серверу можно обратиться с помощью https по адресу домена, указанному в README.md.
- [x] <!--7.28--> Правильно реализовано хранение секретного ключа для создания JWT: 
    - [x] для продакшн сборки он хранится в .env файле — этот файл не добавляется в git;
    - [x] в режиме разработки (при  `process.env.NODE_ENV !== 'production'`) код запускается и работает и без .env файла.

### Хорошие практики
- [x] <!--2.14--> Все роуты подключены в файле `index.js`, который находится в папке `routes`. Оттуда единый роут подключается в файле `app.js`. 
- [x] <!--2.14--> Асинхронные операции реализованы промисами или async/await.
- [x] <!--2.14--> Валидация описана в отдельном модуле.
- [x] <!--2.14--> Настроено логгирование:
    - [x] запросы и ответы записываются в файл `request.log`;
    - [x] ошибки записываются в файл `error.log;`
    - [x] файлы логов не добавляются в репозиторий.
- [x] <!--2.14--> Ошибки обрабатываются централизованным обработчиком.
- [x] <!--2.14--> Централизованный обработчик ошибок описан в отдельном модуле.
- [x] <!--2.14--> API приложения располагается на домене вида: name.zone/api, а не просто name.zone.
    Подходящий адрес: movies-explorer.nomoreparties.co/api.
    Неподходящий адрес: movies-explorer.nomoreparties.co.

### Рекомендации
- [x] <!--0.71--> Для ошибок API созданы классы, расширяющие конструктор `Error`. 
- [x] <!--0.71--> Используется модуль Helmet для установки заголовков, связанных с безопасностью. 
- [ ] <!--0.71--> Конфигурация и константы в отдельных файлах: 
    - [ ] адрес Mongo-сервера, секретный ключ для JWT в режиме разработки, вынесены в отдельный конфигурационный файл;
    - [ ] константы приложения — сообщения ответов и ошибок — вынесены в отдельный файл с константами.
- [x] <!--0.71--> Настроен rate limiter: число запросов с одного IP в единицу времени ограничено. 
- [x] <!--0.71--> Rate limiter сконфигурирован в отдельном файле и импортируется в app.js.
- [x] <!--0.71--> API размещено на отдельном поддомене, например api.movies-explorer.nomoreparties.co. 
- [ ] <!--0.71--> Для имени базы данных используется `moviesdb` 

## Количество баллов: 84

## Доп. Комментарии:
- status `200` передаётся по умолчанию. `res.send` эквивалентен `res.status(200).send`
- Все мидлвары, кроме requestLogger, стоит подключить после limiter. Если limiter отклонит запрос от обработки, то эти мидлвары, подключенные ранее, выполнят лишнюю работу. 
- Нет необходимости генерировать в catch новый объект ошибки и делать проверки на совпадение с текстом сообщений ошибок, созданных в контроллере. Можно сразу сгенерировать ошибку в контроллере, после чего она будет передана в catch. **Напишите в комментарии, пожалуйста, где Вы встретили подобную реализацию с использованием NotValidId**
![image](https://user-images.githubusercontent.com/17438227/147142771-52e4a1eb-fbe0-4d4a-a10d-ba9fd14c4f12.png)

Pancfly commented:
> * Нет необходимости генерировать в catch новый объект ошибки и делать проверки на совпадение с текстом сообщений ошибок, созданных в контроллере. Можно сразу сгенерировать ошибку в контроллере, после чего она будет передана в catch. **Напишите в комментарии, пожалуйста, где Вы встретили подобную реализацию с использованием NotValidId**
>   ![image](https://user-images.githubusercontent.com/17438227/147142771-52e4a1eb-fbe0-4d4a-a10d-ba9fd14c4f12.png)

Добрый вечер. Такой orFail мне посоветовал ревьюер еще на этапе 13й практической работы с ссылкой на [видео](https://www.youtube.com/watch?v=1cIPE4GoMnE). И я использовал его везде в дальнейшем

Pancfly committed:
fix: after review

Add 'ConflictError' - if update email that employed,
Add production 'MongoAdress',
Refactor 'orFail'
Add 'catch' block
